src/main/kotlin/org/wycliffeassociates/otter/common/domain/resourcecontainer/project/IProjectReader.kt

```diff
@@ -0,0 +1,20 @@
+package org.wycliffeassociates.otter.common.domain.resourcecontainer.project
+
+import org.wycliffeassociates.otter.common.collections.tree.Tree
+import org.wycliffeassociates.otter.common.domain.resourcecontainer.ImportResult
+import org.wycliffeassociates.otter.common.domain.resourcecontainer.project.markdown.MarkdownProjectReader
+import org.wycliffeassociates.otter.common.domain.resourcecontainer.project.usfm.UsfmProjectReader
+import org.wycliffeassociates.resourcecontainer.ResourceContainer
+import org.wycliffeassociates.resourcecontainer.entity.Project
+
+interface IProjectReader {
+    fun constructProjectTree(container: ResourceContainer, project: Project): Pair<ImportResult, Tree>
+
+    companion object {
+        fun build(format: String): IProjectReader? = when(format.toLowerCase()) {
+            "text/usfm" -> UsfmProjectReader()
+            "text/markdown" -> MarkdownProjectReader()
+            else -> null
+        }
```

Comment:
l just have a quick question here does it make sense for the interface to have knowledge of it's implementations?should it not only worry about defining its members?  would we not put this somewhere else? I'm just wondering

Reply:
As far as I know, due to JVM classloader optimizations, you can't easily find all implementations programmatically.  I could solve this in one of these ways:

1. *java.util.ServiceLoader*: This requires keeping a separate file listing implementations. Not a real win over this, I think.
2. *Dagger*: I thought about this one a lot, and like it, but I'd have to add a Dagger dependency to this project. I'll do this if you wish.
3. *Kotlin sealed classes*: This is only a small improvement over the current implementation, but it does have the advantage that new implementations would be required to update the list. I'll do this one if you don't want me to add Dagger.

Reply:
@natibekele Could split it into two files, one for the interface and one for a ProjectReaderBuilder, I think that could be an easy compromise?

I may prefer to kick a refactor down the road, maybe we come up with a better option or pattern. I think there is such a thing as a Registry pattern, but I'd have to look into it more.

The interface as it is carries certain dependencies; it's _less_ of an issue because all the parsers are kept in the common repo, though.... In reality we've made the common repo a java-common repo. The following is more for curiosity sake because I'm not advocating a refactor right now:

There are a few things in our common repo that fundamentally disqualify themselves from existing a pure "common" repo as far as Kotlin is concerned. They exist in some of our data classes, and they also exist in the parsers: the IO. File is a java.io.File and we also rely on java.io.BufferedReader, and possibly java.io.InputStream (perhaps implicitly if not explicitly). These already tie us to a java dependency, unless we wanted to create our own File interface, InputStream interface, etc.. where in an actual java-common repo we implement them with java.io.File, etc. This would mean we pretty regularly have to play interface gymnastics when actually connecting our app to the respective target platforms.

The shortest past to alleviating this pain point would be to, at such a time where we created a theoretical js or native repo to port the project to web/native/iOS, etc we theoretically could create a package in those repos named java.io, copy the class names and corresponding method signatures and implement the functions to the java specification based on the respective platform. AFAIK this would in essence be exactly why Oracle sued Google over Android, and Google has lost in the courts twice so far (though, we'd have a better argument that our use is non-commercial, but still legally gray).

As such, adding a dagger dependency does not require an additional dependency we don't already have in common: java.

Reply:
@natibekele @sarabiaj 

The Registry pattern still is limited by the classloader -- an implementation class can't register itself until the CL loads it, and the CL won't bother loading it until it's called for by some other code.

I did look into working around this with sealed classes, but I misunderstood them and they're not a viable solution here, since they require all the impls to be in one file.

As Joe points out, this is a *local* violation of DI, so it's not as big a deal even though it admittedly smells badly.

To mix metaphors, I'd rather not open the can of Dagger worms right now, since Joe is OK kicking this can down the road.  OK to close this?